File: src/server/api-business/repo.js
const { Repo } = require('../models')

/**
 *
 * 获取所属用于的全部Repos
 * @param {User} user
 * @return {List} Repos
 *
 */
async function getUserRepos (user) {
    return (await user.getRepos()).map(repo => ({
        'name': repo.name,
        'descript': repo.descript,
        'attr': repo.attr,
        'user': user.username
    }))
}

/**
 *
 * 查看当前用户是否存在指定的Repo
 * @param {User} user
 * @param {String} 仓库名
 * @return {Boolean} 是否存在该仓库名的Repo
 *
 */
async function existsUserRepo (user, repoName) {
    return !!(await user.getRepos()).find(repo => repo.name === repoName)
}

/**
 *
 * 创建用户仓库
 * @param {User} user
 * @param {String} repoName 用户
 * @param {String} descript 仓库描述
 * @return {Object}
 *
 */
async function createUserRepo (user, repoName, descript) {
    if (await existsUserRepo(user, repoName)) {
        return {
            'result': false,
            'status': 409,
            'message': {
                'reason': 'repo already exists'
            }
        }
    }
    const repo = new Repo({
        user,
        name: repoName,
        descript,
        attr: 0o711
    })
    if (!await repo.create()) {
        return {
            'result': false,
            'status': 500,
            'message': {
                'reason': 'Server Internal Error'
            }
        }
    }
    return {
        'result': true
    }
}

module.exports = {
    getUserRepos,
    createUserRepo
}

File: src/service/bi/service/repo.go
package service

import (
	"net/http"

	"GMOps/src/proto"
	"GMOps/src/service/bi/model"

	"github.com/gin-gonic/gin"
)

type createUserRepoMsg struct {
	Name     string `form:"name"`
	Descript string `form:"descript"`
}

func (s *Service) getUserReposAction(c *gin.Context) {
	user, ok := c.Keys["User"].(*model.User)
	if !ok {
		c.ProtoBuf(http.StatusUnauthorized, &proto.Error{
			ErrorCode: 401,
			Reason:    "Unauthorized",
		})
		return
	}
	repos, err := s.lgc.GetUserRepos(user)
	if err != nil {
		c.ProtoBuf(http.StatusInternalServerError, &proto.Error{
			ErrorCode: 500,
			Reason:    "server internal error",
		})
		return
	}

	ret := &proto.UserRepos{UserId: user.Id.Hex()}
	retRepos := make([]*proto.Repo, 0)
	for _, repo := range repos {
		retRepos = append(retRepos, &proto.Repo{
			Id:       repo.Id.Hex(),
			Name:     repo.Name,
			Descript: repo.Descript,
		})
	}
	ret.Repos = retRepos
	c.ProtoBuf(http.StatusOK, ret)
}

func (s *Service) getUserRepoAction(c *gin.Context) {
	user, ok := c.Keys["User"].(*model.User)
	if !ok {
		c.ProtoBuf(http.StatusUnauthorized, &proto.Error{
			ErrorCode: 401,
			Reason:    "Unauthorized",
		})
		return
	}
	repos, err := s.lgc.GetUserRepos(user)
	if err != nil {
		c.ProtoBuf(http.StatusInternalServerError, &proto.Error{
			ErrorCode: 500,
			Reason:    "server internal error",
		})
		return
	}
	name := c.Param("name")
	for _, repo := range repos {
		if repo.Name == name {
			c.ProtoBuf(http.StatusOK, &proto.Repo{
				Id:       repo.Id.Hex(),
				Name:     repo.Name,
				Descript: repo.Descript,
			})
			return
		}
	}
	c.ProtoBuf(http.StatusNotFound, &proto.Error{
		ErrorCode: 404,
		Reason:    "repo not exist",
	})
}

func (s *Service) createUserRepoAction(c *gin.Context) {
	user, ok := c.Keys["User"].(*model.User)
	if !ok {
		c.ProtoBuf(http.StatusUnauthorized, &proto.Error{
			ErrorCode: 401,
			Reason:    "Unauthorized",
		})
		return
	}
	req := createUserRepoMsg{}
	if err := c.ShouldBind(&req); err != nil {
		s.logger.Error("BI Service createUserRepoAction bind req failed: ", err)
		c.ProtoBuf(http.StatusBadRequest, &proto.Error{
			ErrorCode: 400,
			Reason:    "req body cannot bind",
		})
		return
	}
	if len(req.Name) == 0 {
		c.ProtoBuf(http.StatusBadRequest, &proto.Error{
			ErrorCode: 400,
			Reason:    "req body cannot bind",
		})
		return
	}
	exist, err := s.lgc.ExistUserRepo(user, req.Name)
	if err != nil {
		s.logger.Error("BI Service createUserRepoAction ExistUserRepo failed: ", err)
		c.ProtoBuf(http.StatusInternalServerError, &proto.Error{
			ErrorCode: 500,
			Reason:    "server internal error",
		})
		return
	}
	if exist {
		c.ProtoBuf(http.StatusConflict, &proto.Error{
			ErrorCode: 409,
			Reason:    "repo already existed",
		})
		return
	}
	createdId, err := s.lgc.CreateUserRepo(user, req.Name, req.Descript)
	if err != nil {
		s.logger.Error("BI Service createUserRepoAction ExistUserRepo failed: ", err)
		c.ProtoBuf(http.StatusInternalServerError, &proto.Error{
			ErrorCode: 500,
			Reason:    "server internal error",
		})
		return
	}

	c.ProtoBuf(http.StatusCreated, &proto.Created{Id: createdId})
}
